### DISCUSSION TITLE
how to find elements inside a selector which is only element not elements using sb method?

### DISCUSSION THREAD

**Discussion by SSujitX** [seleniumbase/SeleniumBase#2895] [General]
Here in this code

```
from selenium base import SB

url = 'https://www.sortlist.co.uk/s/360-degree-advertising/united-kingdom-gb'
css_selector = 'li.list-reset'
successfully_filled = 0

with SB(uc=True, incognito=True) as sb:
    try:
        print(">= Opening main URL...")
        sb.driver.uc_open_with_reconnect(url, 3)
        sb.wait_for_element_visible(css_selector, timeout=60)
        
        print(">= Extracting all agency links...")
        all_links = []
        container_element = sb.find_element('div[class="container-lg"]')
        elements = container_element.find_elements('a[class="text-secondary-900 layout-row gap-x-8 cursor-pointer"]')
        for element in elements:
            link = element.get_attribute('href')
            all_links.append(link)
        
        print(f">= Found {len(all_links)} links. Processing each link...")
        
    except Exception as e:
        print(f">= Failed to open main URL or extract links: {e}")

print(f">= Successfully filled {successfully_filled} contact forms.")
```

issue is here,

```
        container_element = sb.find_element('div[class="container-lg"]')
        elements = container_element.find_elements('a[class="text-secondary-900 layout-row gap-x-8 cursor-pointer"]')
```

What is the way to first find an element and then inside the element find all elements? I checked it, and it does not work like this. I found a solution for using a new selector. 

```
            elements = sb.find_elements('div[class="container-lg"] a[class="text-secondary-900 layout-row gap-x-8 cursor-pointer"]')
            for element in elements:
                link = element.get_attribute('href')
                all_links.append(link)

```

**Comment by mdmintz** [comment]
Once you do `WebElement.find_element(*)` or `WebElement.find_elements(*)`, that is using the regular Selenium API, (not the SeleniumBase one), which means no more auto-detecting the selector type, meaning that you have to include the `by` arg, which comes first before the selector. (Eg. `by="css selector"` / `by=By.CSS_SELECTOR`).

So instead of calling `WebElement.find_elements(by="css selector", value=selector)` via the regular Selenium API, you could use the complete selector from `sb.find_elements(selector)`.

**Reply by SSujitX** [reply]
Well, I use the sb.find_elements(selector) method. But I want to mention something. Suppose I want to select a single element and then select multiple elements inside that single element. We can use a table as an example. First, find an element ('table'), then inside the table, find elements ('td') like that. How can I do that?

**Reply by mdmintz** [reply]
Use the full selector of the element(s) you want to find from the beginning, rather than finding one element, and then finding more elements inside of it.

**Comment by luey-punch** [comment]
I have tried these here:

```python
def find_buttons_with_review_ids(self, uniq_id):
    """
    Finds all buttons with both data-review-id and data-href attributes.
    """
    base_path = (
        f"{BASE_XPATH}div/div[1]/div/div/div[2]//button"
    )
    elements = self.find_elements(base_path)
    print(f"rev_elements {elements}")
    reviewers = []
    for element in elements:
        review_id = element.get_attribute("data-review-id")
        print(f"\nreview_id  {review_id}")
        if review_id:
            review_divs = element.find_element("//div")
            print(f"\nreview_divs  {review_divs}")
            review_spans = element.find_element("//span")
            print(f"\nreview_spans  {review_spans}")

    return reviewers
```

I get an error saying **invalid argument: invalid locator**
Adding the 'by' param returns **WebElement.find_elements() got multiple values for argument 'by'**

Even using `element.find_elements("//span","xpath")` still throws an error

**Reply by mdmintz** [reply]
As I mentioned in https://github.com/seleniumbase/SeleniumBase/discussions/2895#discussioncomment-9930193, the `by` arg comes before the `selector` with the original Selenium API.
WebElements use the original Selenium API.
Would be: `element.find_elements("xpath", "//span")`

**Reply by luey-punch** [reply]
Thank you. That fixed it. I am still shocked I used every permutation except this one. I did not see this in the docs. Would be nice if we can update this in docs as it is a common scenario
